### PR TITLE
Add docs for app middleware

### PR DESCRIPTION
**Pull request by mcandeia**
Signed-off-by: Marcos Candeia <marrcooos@gmail.com>

Thank you for helping make the Dapr documentation better!

**Please follow this checklist before submitting:**
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://docs.dapr.io/contributing/contributing-overview/#developer-certificate-of-origin-signing-your-work))
- [x] [Read the contribution guide](https://docs.dapr.io/contributing/docs-contrib/contributing-docs/)
- [x] Commands include options for Linux, MacOS, and Windows within codetabs
- [x] New file and folder names are globally unique
- [x] Page references use shortcodes instead of markdown or URL links
- [x] Images use HTML style and have alternative text
- [x] Places where multiple code/command options are given have codetabs

In addition, please fill out the following to help reviewers understand this pull request:

## Issue reference
Original proposal reference. https://github.com/dapr/dapr/issues/4475 to understand this feature

#2778

### CODE REVIEW

File: daprdocs/content/en/developing-applications/middleware.md [hunk 1]
```diff
@@ -28,19 +28,42 @@ metadata:
...
+        type: middleware.http.oauth2
+      - name: uppercase
+        type: middleware.http.uppercase
 ```
 
 As with other building block components, middleware components are extensible and can be found in the [supported Middleware reference]({{< ref supported-middleware >}}) and in the [components-contrib repo](https://github.com/dapr/components-contrib/tree/master/middleware/http).
 
 {{< button page="supported-middleware" text="See all middleware components">}}
 
+## App Middleware (Preview)
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/middleware.md]: If this is preview then you need to enable this view a preview feature flag? 
You also need to update https://docs.dapr.io/operations/support/support-preview-features/
Add Note heading here rather than in the header.

File: daprdocs/content/en/developing-applications/middleware.md [hunk 2]
```diff
@@ -28,19 +28,42 @@ metadata:
...
+
+Http middleware are executed when invoking Dapr HTTP API directly. There are other types of middlewares that should be applied in the context of app protocol, they are called app middleware and they are also executed when making service-to-service invocations. Valid use cases are token validation in a zero-trust environment or request transformation for a specific app endpoint.
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/middleware.md]: This implies there are two types of middleware, when the feature here is enabling middleware components for service-to-service invocation calls. I would reword this as follows

```suggestion
HTTP middleware components are executed when invoking Dapr HTTP APIs using the `httpPipeline` configuration. You can also use any middleware components when making service-to-service invocation calls for example for token validation in a zero-trust environment, request transformation for a specific app endpoint or apply OAuth policies.
```

File: daprdocs/content/en/developing-applications/middleware.md [hunk 3]
```diff
@@ -28,19 +28,42 @@ metadata:
...
+
+The same set of middlewares that can be applied to http middleware can also be applied to app middleware but instead of using `httpPipeline` you should use `appHttpPipeline`.
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/middleware.md]:
```suggestion
Any middleware component that can be applied to HTTP middleware can also be applied to service-to-service invocation calls as a middleware component using`appHttpPipeline` configuration. The example below adds the `uppercase` middleware component for all outbound calls from the application that this configuration is applied to.
```

File: daprdocs/content/en/developing-applications/middleware.md [hunk 4]
```diff
@@ -28,19 +28,42 @@ metadata:
...
+  name: pipeline
+  namespace: default
+spec:
+  appHttpPipeline:
+    handlers:
+      - name: uppercase
+        type: middleware.http.uppercase
+```
+
+Keep in mind that app middleware are always executed regardless of used Dapr Protocol, when declaring both, app and http middleware, both pipelines are executed.
```
Comment (msfussell) [inline daprdocs/content/en/developing-applications/middleware.md]:
```suggestion
Service-to-service innovation middleware components work for both HTTP and gRPC endpoints.
```

**Comment by msfussell** [comment]
@mcandeia - This feature should be called service-to-service innovation middleware, rather than App middleware. Can you update the diagram title to say this. Also share the diagram somewhere, since this needs to be included into the diagrams deck. @hhunter-ms can help you here. 

Also update the following topics
1) Configuration overview section https://docs.dapr.io/operations/configuration/configuration-overview/#middleware
2) https://docs.dapr.io/operations/support/support-preview-features/ topic.

File: daprdocs/content/en/developing-applications/middleware.md (hunk 3 above)
Comment (msfussell) [reply daprdocs/content/en/developing-applications/middleware.md]: This is a fairly weak example. Could you provide a more compelling example with the existing middleware components?

**Comment by hhunter-ms** [comment]
> @mcandeia - This feature should be called service-to-service innovation middleware, rather than App middleware. Can you update the diagram title to say this. Also share the diagram somewhere, since this needs to be included into the diagrams deck. @hhunter-ms can help you here.
> 
> Also update the following topics
> 
> 1. Configuration overview section https://docs.dapr.io/operations/configuration/configuration-overview/#middleware
> 2. https://docs.dapr.io/operations/support/support-preview-features/ topic.

@mcandeia let me know if you need help adding the diagram to the slide deck

**Comment by msfussell** [comment]
@mcandeia 
1) Update Configuration overview section https://docs.dapr.io/operations/configuration/configuration-overview/#middleware
2) Image for pipeline is wrong and also the arrows on the image you shared are not horizontal. Please work with @hhunter-ms

**Comment by mcandeia** [comment]
> @mcandeia
> 
> 1. Update Configuration overview section https://docs.dapr.io/operations/configuration/configuration-overview/#middleware
> 2. Image for pipeline is wrong and also the arrows on the image you shared are not horizontal. Please work with @hhunter-ms

Sorry folks, I don't know how this image got there

@hhunter-ms definitely yes

**Review by msfussell — APPROVED** [review]
LGTM

File: daprdocs/content/en/developing-applications/middleware.md [hunk 5]
```diff
@@ -5,17 +5,24 @@ linkTitle: "Middleware"
...
+  - /developing-applications/middleware/middleware-overview/
+  - /concepts/middleware-concept/
 ---
 
-Dapr allows custom processing pipelines to be defined by chaining a series of middleware components. A request goes through all defined middleware components before it's routed to user code, and then goes through the defined middleware, in reverse order, before it's returned to the client, as shown in the following diagram.
+Dapr allows custom processing pipelines to be defined by chaining a series of middleware components. There are two places that you can use a middleware pipeline;
 
-<img src="/images/middleware.png" width=800>
+1) Building block APIs - HTTP middleware components are executed when invoking any Dapr HTTP APIs.
+2) Service-to-Service innvocation - HTTP middleware components can also be applied to service-to-service invocation calls.
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/middleware.md]:
```suggestion
2) Service-to-Service invocation - HTTP middleware components are applied to service-to-service invocation calls.
```

File: daprdocs/content/en/developing-applications/middleware.md [hunk 6]
```diff
@@ -5,17 +5,24 @@ linkTitle: "Middleware"
...
+
+## Configuring API middleware pipelines
+
+When launched, a Dapr sidecar constructs a middleware processing pipeline for incoming HTTP calls. By default the pipeline consists of [tracing middleware]({{< ref tracing-overview.md >}}) and CORS middleware. Additional middleware, configured by a Dapr [configuration]({{< ref configuration-concept.md >}}), can be added to the pipeline in the order they are defined. The pipeline applies to all Dapr API endpoints, including state, pub/sub, service invocation, bindings, secrets, configuration, distributed lock and others.
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/middleware.md]:
```suggestion
When launched, a Dapr sidecar constructs a middleware processing pipeline for incoming HTTP calls. By default, the pipeline consists of [tracing middleware]({{< ref tracing-overview.md >}}) and CORS middleware. Additional middleware, configured by a Dapr [configuration]({{< ref configuration-concept.md >}}), can be added to the pipeline in the order they are defined. The pipeline applies to all Dapr API endpoints, including state, pub/sub, service invocation, bindings, secrets, configuration, distributed lock, and others.
```

File: daprdocs/content/en/developing-applications/middleware.md [hunk 7]
```diff
@@ -28,19 +35,43 @@ metadata:
...
 ```
 
-As with other building block components, middleware components are extensible and can be found in the [supported Middleware reference]({{< ref supported-middleware >}}) and in the [components-contrib repo](https://github.com/dapr/components-contrib/tree/master/middleware/http).
+As with other components, middleware components can be found in the [supported Middleware reference]({{< ref supported-middleware >}}) and in the [components-contrib repo](https://github.com/dapr/components-contrib/tree/master/middleware/http).
 
 {{< button page="supported-middleware" text="See all middleware components">}}
 
+## Configuring app middleware pipelines
+
+You can also use any middleware components when making service-to-service invocation calls for example for token validation in a zero-trust environment, a request transformation for a specific app endpoint or to apply OAuth policies.
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/middleware.md]:
```suggestion
You can also use any middleware components when making service-to-service invocation calls. For example, for token validation in a zero-trust environment, a request transformation for a specific app endpoint, or to apply OAuth policies.
```

File: daprdocs/content/en/developing-applications/middleware.md [hunk 8]
```diff
@@ -28,19 +35,43 @@ metadata:
...
+
+Service-to-service invocation middleware components apply to all outgoing calls from Dapr sidecar to the receiving application (service) as shown in the diagram below.
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/middleware.md:52]:
```suggestion
Service-to-service invocation middleware components apply to all outgoing calls from the Dapr sidecar to the receiving application (service), as shown in the diagram below.
```

File: daprdocs/content/en/developing-applications/middleware.md [hunk 9]
```diff
@@ -28,19 +35,43 @@ metadata:
...
+
+<img src="/images/app-middleware.png" width=800>
+
+Any middleware component that can be applied to HTTP middleware can also be applied to service-to-service invocation calls as a middleware component using `appHttpPipeline` configuration. The example below adds the `uppercase` middleware component for all outgoing calls from the Dapr to the application that this configuration is applied to.
```
Comment (hhunter-ms) [inline daprdocs/content/en/developing-applications/middleware.md]:
```suggestion
Any middleware component that can be applied to HTTP middleware can also be applied to service-to-service invocation calls as a middleware component using `appHttpPipeline` configuration. The example below adds the `uppercase` middleware component for all outgoing calls from the Dapr sidecar to the application that this configuration is applied to.
```
